src/tron/tronEngine.ts

```diff
@@ -0,0 +1,1060 @@
+  EdgeLog,
+  EdgeSpendInfo,
+  EdgeTransaction,
+  EdgeWalletInfo,
+  InsufficientFundsError,
+  JsonObject,
+  NoAmountSpecifiedError
+} from 'edge-core-js/types'
+
+import { CurrencyEngine } from '../common/engine'
+import {
+  asyncWaterfall,
+  getDenomInfo,
+  getOtherParams,
+  hexToDecimal,
+  makeMutex,
+  padHex,
+  shuffleArray
+} from '../common/utils'
+import { TronPlugin } from './tronPlugin'
+import {
+  asAccountResources,
+  asBroadcastResponse,
+  asChainParams,
+  asEstimateEnergy,
+  asTransaction,
+  asTRC20Balance,
+  asTRC20Transaction,
+  asTRC20TransactionInfo,
+  asTriggerSmartContract,
+  asTronBlockHeight,
+  asTronQuery,
+  asTRXBalance,
+  asTRXTransferContract,
+  ReferenceBlock,
+  TronAccountResources,
+  TronNetworkFees,
+  TronSettings,
+  TronTxParams,
+  TxQueryCache
+} from './tronTypes'
+import {
+  base58ToHexAddress,
+  encodeTRC20Transfer,
+  hexToBase58Address,
+  TronScan
+} from './tronUtils'
+
+const queryTxMutex = makeMutex()
+
+const ACCOUNT_POLL_MILLISECONDS = 20000
+const BLOCKCHAIN_POLL_MILLISECONDS = 20000
+const TRANSACTION_POLL_MILLISECONDS = 3000
+const NETWORKFEES_POLL_MILLISECONDS = 60 * 10 * 1000
+
+type TronFunction =
+  | 'trx_blockNumber'
+  | 'trx_broadcastTx'
+  | 'trx_chainParams'
+  | 'trx_estimateEnergy'
+  | 'trx_getAccountResource'
+  | 'trx_getBalance'
+  | 'trx_getTransactionInfo'
+  | 'trx_getTransactions'
+
+export class TronEngine extends CurrencyEngine<TronPlugin> {
+  fetchCors: EdgeFetchFunction
+  log: EdgeLog
+  readonly recentBlock: ReferenceBlock
+  accountResources: TronAccountResources
+  networkFees: TronNetworkFees
+  settings: TronSettings
+  accountExistsCache: { [address: string]: boolean }
+  energyEstimateCache: { [addressAndContract: string]: number }
+  tronscan: any
+
+  constructor(
+    currencyPlugin: TronPlugin,
+    walletInfo: EdgeWalletInfo,
+    opts: EdgeCurrencyEngineOptions,
+    fetchCors: EdgeFetchFunction
+  ) {
+    super(currencyPlugin, walletInfo, opts)
+    this.fetchCors = fetchCors
+    this.log = opts.log
+    this.settings = this.currencyInfo.defaultSettings
+      .otherSettings as TronSettings
+    this.recentBlock = {
+      hash: '0',
+      number: 0,
+      timestamp: 0
+    }
+    this.accountResources = {
+      bandwidth: 0,
+      energy: 0
+    }
+    this.networkFees = {
+      createAccountFeeSUN: 100000, // network default
+      bandwidthFeeSUN: 1000, // network default
+      energyFeeSUN: 280 // network default
+    }
+    this.tronscan = new TronScan(this.recentBlock)
+    this.accountExistsCache = {} // Minimize calls to check recipient account resources (existence)
+    this.energyEstimateCache = {} // Minimize calls to check energy estimate
+    this.processTRXTransaction = this.processTRXTransaction.bind(this)
+    this.processTRC20Transaction = this.processTRC20Transaction.bind(this)
+  }
+
+  async fetch(
+    server: string,
+    path: string,
+    opts: Object = {}
+  ): Promise<Object> {
```

Comment:
The return type should be `{ server: string, result: EdgeFetchResponse }`, not `Object.`